README.md

```diff
@@ -1,53 +1,251 @@
+and emit a new state indicating that these records are in flight.
+
+```javascript
+state.length //=> 10
+let record = state.get(7)
+record.isPending //=> true
+record.isResolved //=> false
+record.content //=> null
+```
+
+### Load Horizon
+
+How did it know which records to fetch? The answer is in the
+`loadHorizon` parameter that we passed into the constructor. This
+tells the dataset, that it should keep all records within 10 of the
+current read offset loaded. That's why it fetched the first two
+pages. Now logicially, our dataset looks like this:
+
+> Note: `*` indicates that the record is pending.
+
+
+```
+           Read
+          Offset
+             ┃
+             ┃
+<──────────Load Horizon──────────>
+             ┃
+             ▼
+             ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+              * * * * * * * * * *
+             ◇ ─ ─ ─ ─ ◇─ ─ ─ ─  ┘
+             │         │
+             p0        p1
+```
+
+At some point, the request for the first page resolves. At that point,
+the dataset will emit a new state with the resolved records. That
+state will still contain the pending records as well as the freshly
+loaded.
+
+
+```javascript
+//get a record off the first page
+record = state.get(3);
+record.isResolved //=> true
+record.content //=> 3
+
+//get a record off the second page
+record = state.get(7)
+record.isPending //=> true
+```
+
+Another interesting thing that happened here is that the length of the
+dataset has also changed.
+
+```javascript
+state.length //=> 30
+```
+
+This has to do with the `stats` parameter that is passed into the
+fetch function. This value allows the fetch function to optionally
+specify the total extent of the dataset if that information is
+available. This can be useful when rendering native scrollbars or
+other UI elements that indicate the overall length of a list. If
+`stats` are never updated, then the dataset will just expand
+indefinitely. Now our state looks like this:
+
+> Note `x` indicates that the record is not yet requested
+
+```
+            Read
+           Offset
+              ┃
+              ┃
+<──────────Load Horizon──────────>
+              ┃
+              ▼
+              ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+               0 1 2 3 4 * * * * * xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx│
+              ◇ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ─ ─ ─◇─ ─ ─ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ─ ─ ─
+              │         │         │              │              │
+             p0        p1        p2             p3             p4
+
+```
+
+We have records 0-4, whilst records 5-9 are in flight, and records
+10-25 have yet to be requested.
+
+```javascript
+//from the last page (p4)
+record = state.get(23);
+record.isRequested //=> false
+record.isPending //=> false
+record.content //> null
+```
+
+Let's say we want to move the read head to offset 2 with a call to
+`dataset.setReadOffset(2)`. This will immediately emit a new state that
```

Comment:
should we just do `state.set(2)` since we are able to do `state.get(2)` or is that confusing since we are not setting it to anything? It's not a true definition of getter / setter but i thought i would bring it up.